.github/workflows/native.yml

```diff
@@ -0,0 +1,223 @@
+        sudo apt-get update -o Dir::Etc::sourcelist="/etc/apt/sources.list.d/intel-mkl.list"
+        sudo apt-get install -y --no-install-recommends intel-mkl-64bit-2020.0-088
+    - name: Generate ccache_vars for ccache based on machine
+      shell: bash
+      id: ccache_vars
+      run: |-
+        echo "::set-output name=hash::$(echo ${{ env.ccache_compilercheck }})"
+        echo "::set-output name=timestamp::$(date '+%Y-%m-%dT%H.%M.%S')"
+    - name: Cache-op for build-cache through ccache
+      uses: actions/cache@v2
+      with:
+        path: ${{ env.ccache_dir }}
+        key: ccache-${{ matrix.identifier }}-${{ steps.ccache_vars.outputs.hash }}-${{ github.ref }}-${{ steps.ccache_vars.outputs.timestamp }}
+        restore-keys: |-
+          ccache-${{ matrix.identifier }}-${{ steps.ccache_vars.outputs.hash }}-${{ github.ref }}
+          ccache-${{ matrix.identifier }}-${{ steps.ccache_vars.outputs.hash }}
+          ccache-${{ matrix.identifier }}
+    - name: ccache environment setup
+      run: |-
+        echo "CCACHE_COMPILER_CHECK=${{ env.ccache_compilercheck }}" >> $GITHUB_ENV
+        echo "CCACHE_BASE_DIR=${{ env.ccache_base_dir }}" >> $GITHUB_ENV
+        echo "CCACHE_COMPRESS=${{ env.ccache_compress }}" >> $GITHUB_ENV
+        echo "CCACHE_DIR=${{ env.ccache_dir }}" >> $GITHUB_ENV
+        echo "CCACHE_MAXSIZE=${{ env.ccache_maxsize }}" >> $GITHUB_ENV
+    - name: ccache prolog
+      run: |-
+        ccache -s # Print current cache stats
+        ccache -z # Zero cache entry
+    - name: cmake
+      run: |-
+        mkdir -p build
+        cd build
+        cmake -L .. ${{ matrix.cmake }} ${{ env.ccache_cmake }}
+    - name: Build from source
+      working-directory: build
+      run: make -j2
+    - name: ccache epilog
+      run: 'ccache -s # Print current cache stats'
+    - name: Print Versions
+      working-directory: build
+      run: ./app/bergamot --version
+    - name: Run unit tests
+      working-directory: build
+      run: make test
+      if: ${{ matrix.unittests == 'true' }}
+    - name: Install regression-test framework (BRT)
+      working-directory: bergamot-translator-tests
+      run: make install
+    - name: Run regression-tests (BRT)
+      working-directory: bergamot-translator-tests
+      id: brt_run
+      run: MARIAN=../build ./run_brt.sh ${{ matrix.brt_tags }}
+    - name: Print logs of unsuccessful BRTs
+      working-directory: bergamot-translator-tests
+      run: grep "tests.*.sh" previous.log  | sed 's/^\s*-\s*//' | xargs -I% tail -n100 -v %.log
+      if: ${{ always() && steps.brt_run.outcome == 'failure' }}
+    - name: Upload regression-tests artifacts
+      uses: actions/upload-artifact@v2
+      if: ${{ always() && steps.brt_run.outcome != 'skipped' }}
+      with:
+        name: brt-${{ matrix.identifier }}
+        path: |-
+          bergamot-translator-tests/**/*.expected
+          bergamot-translator-tests/**/*.log
+          bergamot-translator-tests/**/*.out
+  mac:
```

Comment:
Do we prefer the Ubuntu and MacOS jobs to be separate? 
 
If we wanted to reduce some of the duplication it seems we could use either: 
```yaml
if: runner.os == 'Linux'
if: runner.os == 'macOS' 
```
to break run Linux
 - Linux dependencies
 - MKL
 
and MacOS
 - MacOS dependencies
 - BLAS paths

specific portions.

Reply:
1. I don't feel positive about the if in the job steps code and a runner-os switch. We put such an `if` in the workflow, it makes the editing workflow a bit more complex than it needs to be (For example, I'm now thinking of tweaking the matrix combos vs adding another if to achieve the same thing). 

2. An ideal solution that contains things within the YAML IMO will be to do `${{setup-deps}}` at each matrix include-entry, and have it evaluated in a setup-deps step. However, this will create duplication in the matrix. I'm trying to add as less programming to the YAML as possible keeping it simple and stupid.

3. Another nice way I see people doing this is replace this is with a `bash setup.sh <build-type>` and run a `bash setup.sh ${build-type}}` loading build-type from matrix, where `build-type=(['ubuntu', 'mac'] x ['minimal', 'full'])`, which will internally makes the switch. This way we only use the matrix (less duplicity because build-type is a string) and no ifs to do the same thing as the matrix within job steps. The downside here is that things are no longer contained within the YAML.

Among the possibilities, I think the current state is the sweet spot (Process is contained in YAML, and the workflow is simple and stupid at the cost of some duplication).

Reply:
I'm happy for Ubuntu and MacOS to exist in distinct matrix runs.

3 is a desirable solution if we imagine people are going to run `setup.sh` locally, but I don't imagine they will as it's purely setting up the environment. Let's just stay as is?